smtpserver/server_test.go

```diff
@@ -1767,3 +1767,63 @@ func TestFutureRelease(t *testing.T) {
 	test(" HOLDFOR=1 HOLDUNTIL="+time.Now().Add(time.Hour).UTC().Format(time.RFC3339), "501")                        // Duplicate.
 }
+
+// Test SMTPUTF8
+func TestSMTPUTF8(t *testing.T) {
+	ts := newTestServer(t, filepath.FromSlash("../testdata/smtp/mox.conf"), dns.MockResolver{})
+	defer ts.close()
+
+	ts.user = "mjl@mox.example"
+	ts.pass = password0
+	ts.submission = true
+
+	test := func(mailFrom string, rcptTo string, headerValue string, filename string, clientSmtputf8 bool, expectedSmtputf8 bool, expErr *smtpclient.Error) {
+		t.Helper()
+
+		ts.run(func(_ error, client *smtpclient.Client) {
+			t.Helper()
+			msg := strings.ReplaceAll(fmt.Sprintf(`From: <%s>
+To: <%s>
+Subject: test
+X-Custom-Test-Header: %s
+MIME-Version: 1.0
+Content-type: multipart/mixed; boundary="simple boundary"
+
+--simple boundary
+Content-Type: text/plain; charset=UTF-8;
+Content-Disposition: attachment; filename="%s"
+Content-Transfer-Encoding: base64
+
+QW4gYXR0YWNoZWQgdGV4dCBmaWxlLg==
+
+--simple boundary--
+`, mailFrom, rcptTo, headerValue, filename), "\n", "\r\n")
+
+			err := client.Deliver(ctxbg, mailFrom, rcptTo, int64(len(msg)), strings.NewReader(msg), false, clientSmtputf8, false)
+			var cerr smtpclient.Error
+			if expErr == nil && err != nil || expErr != nil && (err == nil || !errors.As(err, &cerr) || cerr.Code != expErr.Code || cerr.Secode != expErr.Secode) {
+				t.Fatalf("got err %#v, expected %#v", err, expErr)
+			}
+			if err != nil {
+				return
+			}
+
+			msgs, _ := queue.List(ctxbg, queue.Filter{})
+			queuedMsg := msgs[len(msgs)-1]
+			if queuedMsg.SMTPUTF8 != expectedSmtputf8 {
+				t.Fatalf("[%s / %s / %s / %s] got SMTPUTF8 %t, expected %t", mailFrom, rcptTo, headerValue, filename, queuedMsg.SMTPUTF8, expectedSmtputf8)
+			}
+		})
+	}
+
+	test(`mjl@mox.example`, `remote@example.org`, "header-ascii", "ascii.txt", false, false, nil)
+	test(`mjl@mox.example`, `remote@example.org`, "header-ascii", "ascii.txt", true, false, nil)
+	test(`mjl@mox.example`, `🙂@example.org`, "header-ascii", "ascii.txt", true, true, nil)
+	test(`mjl@mox.example`, `🙂@example.org`, "header-ascii", "ascii.txt", false, true, &smtpclient.Error{Permanent: true, Code: smtp.C553BadMailbox, Secode: smtp.SeMsg6NonASCIIAddrNotPermitted7})
+	test(`Ω@mox.example`, `remote@example.org`, "header-ascii", "ascii.txt", true, true, nil)
+	test(`Ω@mox.example`, `remote@example.org`, "header-ascii", "ascii.txt", false, true, &smtpclient.Error{Permanent: true, Code: smtp.C550MailboxUnavail, Secode: smtp.SeMsg6NonASCIIAddrNotPermitted7})
+	test(`mjl@mox.example`, `remote@example.org`, "header-utf8-😍", "ascii.txt", true, true, nil)
+	test(`mjl@mox.example`, `remote@example.org`, "header-ascii", "utf8-🫠️.txt", true, true, nil)
+	test(`Ω@mox.example`, `🙂@example.org`, "header-utf8-😍", "utf8-🫠️.txt", true, true, nil)
+	test(`mjl@mox.example`, `remote@idn-🌏️.org`, "header-ascii", "ascii.txt", true, true, nil)
```

Comment:
Not sure about this one.

The utf8 character is only in the recipient domain, not the localpart. If I'm not wrong, the SMTPUTF8 extension is required **only** if the domain is not _translated_ to punycode. So this test should fail?

Reply:
well, the smtpclient will send that literal recipient address to the smtpserver, with the utf8 extension, so smtputf8 and msgsmtputf8 will be true. the rcptto address is also included in the test message in the To header.
i think the test should be for a rcptto address with punycode a-labels (and also use a subdomain under example.org, not a domain that could actually exist, e.g. xn--vg8h.example.org for 🌏️.example.org).